### ISSUE TITLE
Colocating form logic

### ISSUE THREAD

**Issue by deoqc** [erikras/redux-form#1577]
tl;dr: With the major rewrite of the package (aka v6), is the opportunity to **colocate** most of the logic of each form together - which means they should all be in `reduxForm` and none in the `reducer`.

---

### Normalization

The normalization, in v5, was in the reducer. Why not, for example, to enhance `onChange` so that each value was transformed before being dispatched? 

Something like:
```Javascript
// generic onChange enhancer: the enhancer should receive a value and return a value
const enhanceOnChange = (enhancer) => (onChange) => (value) => (
  onChange(enhancer(value))
)

// have normalizeField that takes a value and returns it normalized
// new onChange, with normalizeField
onChange = enhanceOnChange(normalizeField)(onChange)
```

### Sync validation

The decision for moving the sync validation to the reducer seems wrong to me. Even if the decision is to keep the error in the store - I would prefer to get the errors as a selector, since is derived data - is perfectly possible to `enhanceOnChange` a second time to dispatch the error after a possible normalization.

```Javascript
// normalizeField already have the signature correct, and don't need any help from redux-form
// for validateField is necessary an intermediate generic errorEnhancer, 
// and a specify dispatchError provided by redux-form for this field

// dispatchError: provided by redux-form to dispatch the error for this specific fields
// validateField: is the validation provided by the user for this specific field

const errorEnhancerHelper = (dispatchError, validateField) => (value) => {
  dispatchError(validateField(value)) // compute and dispatch error
  return value
}

const errorEnhancer = errorEnhancerHelper(dispatchError, validateField)

const composedEnhancer = compose( // 'lodash/fp/compose'
  errorEnhancer,
  normalizeField
)

// enhanced onChange
onChange = enhanceOnChange(composedEnhancer)(onChange)
```
Also, if the decision is to not have the errors in redux store anymore, could have minor change in the api. Substitute the `dispatchError` for `assignErrorToField`.

---

ps: The important point is the **colocation**, the enhancer is implementation detail. But maybe could incorporate `onChange` enhancer for the library also so user could easily write our custom logic...

ps: I don't have deep knowledge of 'redux-form' v5, and almost none of v6. I am certainly overlooking/oversimplifying lots of things...

**Comment by erikras** [comment]
Yes, a lot of thought was put into the `v6` rewrite.

I liked the idea of the normalization being in the reducer in v5 because it guaranteed that any `dispatch(change(form, name, value))` from anywhere in the application would have normalization applied to it. However, that benefit was outweighed by exactly the argument you are proposing, and in `v6`, the normalization is pretty much just an enhanced `onChange`. See [Value Lifecycle](http://redux-form.com/6.0.0-rc.5/docs/ValueLifecycle.md/).

I tried really, really, really hard to keep the sync errors out of the Redux store, but when I finally succumbed, it made _so_ many complex bugs disappear. It was really complicated to rerender only the field whose sync error had changed when the sync errors were coming down from the outer form component via `context`.

**Comment by deoqc** [comment]
In v5, I had to implement normalizing the deep fields because *IFAK* redux-form didn't support it. And I almost switched to implementing all normalizing by myself, because I really value colocating the logic together - It was more maintainable to have complex code together than simple code spread apart.

**Couldn't the errors be dispatched in the Value Lifecycle also? (just beside 'CHANGE action')** 

It will be a nightmare to have multiple forms across an app having this fundamental logic in the far away reducer, which I wouldn't like to touch. 


*ps: Since v6 have `Field`, maybe this is the proper place to declare the sync validation and the normalization. It would be further colocating the logic in the `Field`, instead of passing down from top-level `reduxForm`. But I don't wanna to push this idea now =).*

**Comment by erikras** [comment]
> Couldn't the errors be dispatched in the Value Lifecycle also? (just beside 'CHANGE action')

Yes, this has been considered. It is certainly another option.

> It will be a nightmare to have multiple forms across an app having this fundamental logic in the far away reducer, which I wouldn't like to touch.

I'm not sure what "fundamental logic" you are referring to here.

**Comment by deoqc** [comment]
Well, maybe "fundamental logic" should be refrased to "specific logic".  I was referring to the sync validation in v6 (but could be to normalization in v5 also).

IMHO it makes lot a difference to have the logic together - and not separate in component reduxForm and reducer.

Thanks for hearing my concerns and for this great package, and hope to have some way around the sync validation in the reducer - having it back in reduxForm or in the new Field would be great.

**Comment by erikras** [comment]
> Thanks for hearing my concerns and for this great package

Thanks for voicing them. ❤️

> hope to have some way around the sync validation in the reducer

Sync validation isn't so much _IN_ the reducer – [It was once! 👻](http://redux-form.com/6.0.0-alpha.4/examples/syncValidation/) – as it is "using the reducer to store sync validation errors as a way of communicating between the Form and the Field(s)".

I've enjoyed your respectful challenge to my decisions. All of OSS would be stronger if more people were like you. 👍

**Comment by deoqc** [comment]
Looks like the [sync validation in reduxForm](http://redux-form.com/6.0.1/examples/syncValidation/) is back! Great!
